File: tools/promote-poster.py
#!/usr/bin/env python3
"""promote-poster.py — encode the producer's terminal raster into the shipped PNG.

The pixels come from tools/promote-poster.js under jsc — the repository's own
substrate kernel, not a second renderer. This script is a lossless container step
and a receipt printer: it never computes a pixel. tools/test-poster.py decodes the
result and compares it back to the producer, so the encoder cannot become an
authority over the image.

Run twice; the output SHA-256 must be identical (fixed zlib level, filter 0, no
ancillary chunks — tools/pngout.py).
"""
import base64, hashlib, os, subprocess, sys

sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))
from pngout import write_png

JSC = "/System/Library/Frameworks/JavaScriptCore.framework/Versions/A/Helpers/jsc"
ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
OUT = os.path.join(ROOT, "assets/plate/surface-terminal.png")
ENCODER = "tools/pngout.py write_png(alpha=True) — PNG color type 6, 8-bit, filter 0 every row, zlib level 9, no ancillary chunks"


def produce(root=ROOT, manifest="js/manifest.js"):
    """(provenance dict, RGBA bytes) from the substrate kernel under jsc."""
    run = subprocess.run([JSC, os.path.join(root, manifest),
                          os.path.join(root, "js/substrate.js"),
                          os.path.join(ROOT, "tools/promote-poster.js")],
                         capture_output=True, text=True)
    if run.returncode != 0:
        raise SystemExit("producer failed: " + run.stderr.strip())
    import json
    prov, rgba = None, None
    for line in run.stdout.splitlines():
        if line.startswith("PROVENANCE "):
            prov = json.loads(line[len("PROVENANCE "):])
        elif line.startswith("RGBA "):
            rgba = base64.b64decode(line[len("RGBA "):])
    if prov is None or rgba is None:
        raise SystemExit("producer emitted no raster")
    if len(rgba) != prov["rgba_bytes"]:
        raise SystemExit("raster length disagrees with the provenance line")
    return prov, rgba


def main() -> int:
    prov, rgba = produce()
    w, h = (int(v) for v in prov["raster"].split("x"))
    write_png(OUT, w, h, [rgba[y * w * 4:(y + 1) * w * 4] for y in range(h)], alpha=True)
    png = open(OUT, "rb").read()
    print("PROMOTED  assets/plate/surface-terminal.png")
    for key in ("plate", "reference_frame", "raster", "bin_scale", "checksum", "epoch",
                "terminal_step", "deposits", "iterations", "max_density", "density_hash_fnv1a"):
        print(f"  {key:22s} {prov[key]}")
    print(f"  {'params':22s} {prov['params']}")
    print(f"  {'terminal RGBA sha256':22s} {hashlib.sha256(rgba).hexdigest()}")
    print(f"  {'poster asset sha256':22s} {hashlib.sha256(png).hexdigest()}")
    print(f"  {'bytes':22s} {len(png):,}")
    print(f"  {'encoder':22s} {ENCODER}")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: backup/instrument-v1/tools/promote-poster.js
/* tools/promote-poster.js — the offline producer for the promoted terminal exposure.

   Classification: DERIVED. It is not a second renderer: it runs js/substrate.js —
   the same kernel the browser runs — over the canonical manifest, develops plate 0
   to its terminal state D_N, and tonemaps it with the same tonemapInto the page
   uses. Since §8.1 every recurrence is dsin/dcos/datan2, so D_N and the raster R_N
   are engine-invariant: jsc and the browser produce identical bytes.

   Emits one PROVENANCE line then the terminal RGBA raster, base64, one line.
   Encoding to PNG is a lossless container step done by tools/promote-poster.py;
   tools/test-poster.py decodes the shipped file and compares pixels back to this
   producer, so the container never becomes an authority over the image.

   Run: jsc js/manifest.js js/substrate.js tools/promote-poster.js */
"use strict";
var S = globalThis.CytherSubstrate, CM = globalThis.CytherManifest;

/* The reference frame is a promotion decision, fixed here so the command takes no
   arguments and the receipt is a pure function of the repository. 1200x600 is a
   real viewport frame whose bin raster is exactly BIN_TGT (720,000 cells) — the
   most the runtime ever develops — and 2:1 is the widest aspect the page presents
   it at, so a cover-fit at any landscape aspect in [1,2] is the exact crop the
   plate for that viewport would show. */
var REF = { plate: 0, W: 1200, H: 600 };

var B64 = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";
function base64(u8) {
  var parts = [], chunk = [], i = 0, n = u8.length;
  for (; i + 2 < n; i += 3) {
    var v = (u8[i] << 16) | (u8[i + 1] << 8) | u8[i + 2];
    chunk.push(B64[(v >> 18) & 63] + B64[(v >> 12) & 63] + B64[(v >> 6) & 63] + B64[v & 63]);
    if (chunk.length === 8192) { parts.push(chunk.join("")); chunk = []; }
  }
  var rest = n - i;
  if (rest === 1) { var a = u8[i] << 16; chunk.push(B64[(a >> 18) & 63] + B64[(a >> 12) & 63] + "=="); }
  else if (rest === 2) { var b = (u8[i] << 16) | (u8[i + 1] << 8); chunk.push(B64[(b >> 18) & 63] + B64[(b >> 12) & 63] + B64[(b >> 6) & 63] + "="); }
  parts.push(chunk.join(""));
  return parts.join("");
}

var cam = S.deriveAnchors(CM.CANON);
var frame = S.frameFor(cam.bounds, REF.W, REF.H);
var st = S.plateState(CM.CANON, REF.plate, cam.ANCH[REF.plate], frame);
while (!st.done) S.developStep(st, CM.CANON);
/* Uint8ClampedArray, not Uint8Array: ImageData's round-half-even clamping is part
   of the exposure law, and the browser's raster must round identically. */
var rgba = new Uint8ClampedArray(frame.bw * frame.bh * 4);
S.tonemapInto(st, rgba);

print("PROVENANCE " + JSON.stringify({
  producer: "tools/promote-poster.js",
  plate: REF.plate,
  reference_frame: REF.W + "x" + REF.H,
  raster: frame.bw + "x" + frame.bh,
  bin_scale: frame.sc,
  params: CM.CANON,
  checksum: CM.CHECKSUM,
  epoch: CM.MANIFEST.epoch,
  terminal_step: st.k,
  deposits: st.dep,
  iterations: st.it,
  max_density: st.maxT,
  density_hash_fnv1a: S.stateHash(st),
  rgba_bytes: rgba.length
}));
print("RGBA " + base64(rgba));
